### PR TITLE
cannonlake: use low power ring osc as default dsp clock

### PR DESCRIPTION
**Pull request by mmaka1**
Running on 120MHz Low Power Ring Oscillator enables more
aggressive power saving. Therefore this should be a default
options for all Cannonlake configurations that may run
on a single dsp core and does not need more than 120 MCPS.

Signed-off-by: Marcin Maka <marcin.maka@linux.intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
@mmaka1 for my education, with the LPRO set as default clk source, if we run multi-pipeline tests, would the clock source get switched dynamically?

**Comment by xiulipan** [comment]
@macchian It seems with this PR we will hit issue https://github.com/thesofproject/sof/issues/2176 more easily.

**Comment by mmaka1** [comment]
> @mmaka1 for my education, with the LPRO set as default clk source, if we run multi-pipeline tests, would the clock source get switched dynamically?

No, there is no dynamic switch implemented. There are some APIs in the clk module but there is no logic that would monitor the load and throttle the clock appropriately yet.

However, this is still recommended to go with the LP RO clocks (and keep it as the default power policy along with low power sequencer and others) unless specific topology requires more then 120 mcps. Of course, somebody may opt for running at full speed until the dynamic clk switching is implemented, however this may hide a lot of performance/timing issues that should be debugged and fixed. I'd expect some bugs repro rate to increase with this setting which might be helpful (like the one #2176 mentioned).

So, do we have any released-and-used topology that **really** requires more than 120? Which of the tests you mentioned need more?

File: src/arch/xtensa/configs/cannonlake_defconfig [hunk 1]
```diff
@@ -3,6 +3,7 @@ CONFIG_CAVS_DMIC=y
 CONFIG_CAVS_SSP=y
 CONFIG_CAVS_ALH=y
 CONFIG_CAVS_LPS=y
+CONFIG_CAVS_LPRO=y
```
Comment (jajanusz) [inline src/arch/xtensa/configs/cannonlake_defconfig]: Can you change this line also:
https://github.com/thesofproject/sof/blob/e9bec4a559208ee8ae6b44ac24bcd8943bd0f4ee/src/platform/Kconfig#L224
It's better to have defaults in defconfigs than in Kconfig files

**Comment by mmaka1** [comment]
> Hmm, change looks good, but I wonder if wider testing is needed. CI device test shows error in both CML and CFL systems.

@kv2019i Yes, I noticed the failures, but I am not able to measure if the lower mcps budget has any real impact (I had not expected any) right now since the DMIC capture using failing topologies, no clock lowered yet, is very unstable at the moment (#2202, #2176) so there is no stable base to run any tests.

**Comment by kv2019i** [comment]
@mmaka1 wrote:
> @kv2019i Yes, I noticed the failures, but I am not able to measure if the lower mcps budget has any real impact (I had not expected any) right now since the DMIC capture using failing topologies, no clock lowered yet, is very unstable at the moment (#2202, #2176) so there is no stable base to run 

Ack on that. It seems @singalsu found an older commit that has triggered #2176 , so hopefully we get a fix soon.

**Comment by jajanusz** [comment]
@mmaka1 Please merge it only if CIs are green. If there are issues discrovered by making 120mhz as default clock then they need to be handled.

**Comment by mmaka1** [comment]
@jajanusz I am aware of #2208, but not sure if it is covered by CI. "Green" status may not include that one.

**Comment by singalsu** [comment]
Playback via SRC with 44.1 kHz content when FW is built with gcc would fail but there does not seem to be use of pipe-pcm-media.m4 for topologies for CNL and after.

**Comment by jajanusz** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@mmaka1 only 2 legit looking CI failure now, I guess this will be ready for merge soon ?

**Comment by mmaka1** [comment]
@lgirdwood Those two cases are failing because of dmic fifo overruns and it seems that some low latency cycles takes longer than 1ms with 120MHz clock which is unexpected with that simple topo. Investigating it right now. Since I am not able to repro on my cfl machine, adding more logs to schedulers/timers logs (will be useful anyway) and moving others, less relevant, to verbose level. So more updates are coming to run more tests on CI machines. **Pls do not merge until [WIP] tag is removed!**

**Comment by singalsu** [comment]
Is there difference with DMIC FIFO overruns with xt-xcc and gcc builds? The IIR computation is much faster in xt-xcc version. The DMIC capture pipes contain the IIR DC blocker filters.

**Comment by lgirdwood** [comment]
@singalsu maybe another reduction in filter leng for GCC build. It should just be functional with minimal filter length. Performance need to be great as production will all use xcc, but some testers/devs will use GCC.

**Comment by mmaka1** [comment]
@singalsu Yes, eq iir may be the heaviest contributor here however it'd be nice to have simple profiling info available instantly from the traces in order to improve debugability, so I am instrumenting schedulers code now. Even if compiled with gcc, would it need >120MCPS just for two instances?

**Comment by singalsu** [comment]
The used 2nd order IIR (single biquad) is the lowest possible configuration. Though it should be possible to optimize a bit the gcc version. I've put only effort to xcc version since the speed of gcc version has not before been an issue.

Omitting IIR from gcc build is not really an option because then we'd get the subjective quality issues in capture with "thump" sounds into the beginning.

I'll measure the copy() execution times on WHL next. Do you know what is the MHz of WHL DSP to know the MCPS? And does this PR impact also WHL so I could measure @ 120 MHz too?

**Comment by singalsu** [comment]
On WHL with default clock and GCC build for DMIC device hw:0,6 and hw:0,7 capture there's 23% load from filters.

-48 kHz 4ch S32_LE IIR needs 167 us every 1000 us, 17% load

- 16 kHz 4ch S32_LE IIR needs 60us , 6% load

As xt-xcc build, 7% load

-48 kHz 4ch S32_LE IIR needs 50 us, 5% load

- 16 kHz 4ch S32_LE IIR needs 22us , 2% load

Running the pipes simultaneously does not seem to increase the processing times per component, there's  maybe even some tiny drop. I did this by looking the trace and rounded to integer microseconds largest looking processing times. Now that trace format is stabilized I should parse propoer min/max/avg statistics from entire run.

**Comment by singalsu** [comment]
I tested with this PR merged GCC, total 76%

- 48 kHz 4ch S32_LE IIR needs 557 us, 56% load

- 16 kHz 4ch S32_LE IIR needs 199 us, 20% load

With xt-xcc build, total 23%

- 48 kHz 4ch S32_LE IIR needs 160 us, 16% load

- 16 kHz 4ch S32_LE IIR needs 65 us, 7% load

The gcc build gets really close to realtime limit unfortunately. I'll see if I can do some quick optimization for generic C IIR core.

**Comment by singalsu** [comment]
There's no gain from going to full pointer arithmetic (much like the HiFi3 version) despite a lot simpler looking IIR core. The compiler handles really well arrays it seems. The 64 bit multiplication is likely the killer for performance since gcc doesn't use the native xtensa 32 x 24 -> 48 bit multiply. 

Do you mind if I ask is applying the 120 MHz clock needed for gcc build, could it be a feature for xt-xcc build only? Also are there plans to change clock automatically based on pipelines activity on DSP?

**Comment by jajanusz** [comment]
@singalsu You may have secondary config for GCC and name it defconfig_cannonlake_gcc, but CI should by default test xcc with 120mhz as it's what will be running on most of machines

**Comment by singalsu** [comment]
@jajanusz OK, in that case I don't the the IIR processing load could cause issues.

**Comment by lgirdwood** [comment]
@mmaka1 any ETA for merging ?

**Comment by mmaka1** [comment]
> @mmaka1 any ETA for merging ?

How should we deal with the gcc builds that do not fit into 120 MCPS budget (tests are failing)? Do you want the gcc builds to use 400 MHz clock as the default one?

**Comment by lgirdwood** [comment]
@mmaka1 yes, I would still allow GCC to build with this option enabled, but we should have a big warning in the ring OSC Kconfig menu to state that some GCC compiled code may not run to scheduling deadlines with this option enabled. Fwiw, simple pipelines do work with volume with 50MHz clock on BYT and GCC.

**Comment by mmaka1** [comment]
@lgirdwood OK, so wrt to the defconfig-s used by CI builds, do you want to follow BYT/SUE way and fork cannonlake_xcc_defconfig adding LPRO=Y only to that one?

**Comment by lgirdwood** [comment]
@mmaka1 I think it's best to say that XCC is the preferred compiler when the xtensa architecture is selected in Kconfig and that performance is not guaranteed with GCC on xtensa. I dont want to have xcc or gcc defconfigs, but have a defconfig only for the preferred compiler (XCC). Yes we can have GCC detection in our source files to lower capabilities, but my expectation is that all code should compile with both compilers.

File: scripts/xtensa-build-all.sh [hunk 2]
```diff
@@ -281,10 +281,10 @@ do
 		PATH=$pwd/../$HOST/bin:$OLDPATH
 		COMPILER="gcc"
 
-		if [ $j == "byt" ] || [ $j == "cht" ] || [ $j == "sue" ]
-		then
-			DEFCONFIG_PATCH="_gcc"
-		fi
+		case $j in
+			byt|cht|cnl|sue) DEFCONFIG_PATCH="_gcc"
```
Comment (lgirdwood) [inline scripts/xtensa-build-all.sh]: Should this be in previous patch ?

Comment (mmaka1) [reply scripts/xtensa-build-all.sh]: @lgirdwood I believe the latest version is ok. The first patch just swaps gcc/xcc extensions, the second one adds cnl as an exception for gcc builds as well.

**Comment by lgirdwood** [comment]
@mmaka1 @zrombel btw, it looks like internal CI results are partial for this PR ? Do we need to rerun ?

**Comment by mmaka1** [comment]
> @mmaka1 @zrombel btw, it looks like internal CI results are partial for this PR ? Do we need to rerun ?

Re-run of internal CI passed.

**Pull request merged**